Automattic/wp-calypso
# A4A: add marketplace type toggle to allow referrals

Resolves https://github.com/Automattic/jetpack-genesis/issues/356

## Proposed Changes

This PR aims to add a switch of what type of Marketplace we are in. Currently, there are 2 types, `referral` and `regular`. I've  tried to be less intrusive into current logic, so it's just adds `type` to shopping cart and to additional context provider, which will be used in the future.
Shopping Cart relies different session storages, depending in which Marketplace type we are in.

I think it looks quite smooth but I will leave it for @jeffgolenski and @madebynoam to verify.
**Note**: It's only general change for now, so cards and checkout are not yet updated.

https://github.com/Automattic/wp-calypso/assets/60262784/d666a1c8-0425-40d5-836d-228baaa5eb37

## Why are these changes being made?


## Testing Instructions



- Navigate to `/marketplace/products/` and check the toggle and it's behavior.
- Navigate to `/marketplace/products?purchase_type=referral`, you should be immediately in referral mode.
- Check that with flag disabled `flags=-a4a-automated-referrals` Marketplace acts as before.

## Pre-merge Checklist
